### DISCUSSION TITLE
Haskell E-Graph library: Overeasy

### DISCUSSION THREAD

**Discussion by ejconlon** [egraphs-good/egg#161] [Show and tell]
Hello all fellow egg-thusiasts,

I read the egg paper in a reading group last fall and fell in love with e-graphs! So much so that I went and implemented one in Haskell: https://github.com/ejconlon/overeasy . (Having egg as a reference was pretty valuable.)

Though there are some (really) messy parts, it seems to work, passing the unit and property tests I have. Two cool things to note are 1) that it's a composition of immutable data structures, so you can use it seamlessly in a backtracking search, and 2) it supports arbitrary traversable functors as terms (meaning you bring whatever AST you want).

Please reach out if you are interested in using it or collaborating, or if you're just into the intersection of e-graphs, logic, SAT/SMT solving and functional programming.

Thanks,
Eric

**Comment by mwillsey** [comment]
Awesome work! I'm glad egg was valuable as a reference :)
